tools/clang/lib/SPIRV/GlPerVertex.cpp

```diff
@@ -0,0 +1,751 @@
+
+  switch (semantic->GetKind()) {
+  case hlsl::Semantic::Kind::ClipDistance:
+    typeMap = asInput ? &inClipType : &outClipType;
+    break;
+  case hlsl::Semantic::Kind::CullDistance:
+    typeMap = asInput ? &inCullType : &outCullType;
+    isCull = true;
+    break;
+  default:
+    // Annotated with something other than SV_ClipDistance or SV_CullDistance.
+    // We don't care about such cases.
+    return true;
+  }
+
+  // Parameters marked as inout has reference type.
+  if (baseType->isReferenceType())
+    baseType = baseType->getPointeeType();
+
+  if (baseType->isFloatingType() || hlsl::IsHLSLVecType(baseType)) {
+    (*typeMap)[semanticIndex] = baseType;
+    return true;
+  }
+
+  if (baseType->isConstantArrayType()) {
+    if (shaderModel.IsHS() || shaderModel.IsDS() || shaderModel.IsGS()) {
+      // Ignore the outermost arrayness and check the inner type to be
+      // (vector of) floats
+
+      const auto *arrayType = astContext.getAsConstantArrayType(baseType);
+
+      // TODO: handle extra large array size?
+      if (*blockArraySize !=
+          static_cast<uint32_t>(arrayType->getSize().getZExtValue())) {
+        emitError("inconsistent array size for shader %select{output|input}0 "
+                  "variable '%1'",
+                  decl->getLocStart())
+            << asInput << decl->getName();
+        return false;
+      }
+
+      const QualType elemType = arrayType->getElementType();
+
+      if (elemType->isFloatingType() || hlsl::IsHLSLVecType(elemType)) {
+        (*typeMap)[semanticIndex] = elemType;
+        return true;
+      }
+
+      emitError("elements for %select{SV_ClipDistance|SV_CullDistance}0 "
+                "variable '%1' must be (vector of) floats",
+                decl->getLocStart())
+          << isCull << decl->getName();
+      return false;
+    }
+
+    emitError("%select{SV_ClipDistance|SV_CullDistance}0 variable '%1' not "
+              "allowed to be of array type",
+              decl->getLocStart())
+        << isCull << decl->getName();
+    return false;
+  }
+
+  emitError("incorrect type for %select{SV_ClipDistance|SV_CullDistance}0 "
+            "variable '%1'",
+            decl->getLocStart())
+      << isCull << decl->getName();
+  return false;
+}
+
+void GlPerVertex::calculateClipCullDistanceArraySize() {
+  // Updates the offset map and array size for the given input/output
+  // SV_ClipDistance/SV_CullDistance.
+  const auto updateSizeAndOffset = [](const SemanticIndexToTypeMap &typeMap,
+                                      SemanticIndexToArrayOffsetMap *offsetMap,
+                                      uint32_t *totalSize) {
+    // If no usage of SV_ClipDistance/SV_CullDistance was recorded,just
+    // return. This will keep the size defaulted to 1.
+    if (typeMap.empty())
+      return;
+
+    *totalSize = 0;
+
+    // TODO: We are doing extra queries here. Should also report error if the
+    // index is > 9.
+    for (uint32_t i = 0; i < 10; ++i) {
```

Comment:
it seems like an unnecessary limitation. If DenseMap is an ordered map, you can use `typeMap.rbegin()` to find the entry with the largest `i`. If it's not an ordered map, should we use an ordered map? (given the number of Clip/Cull vars, it shouldn't be expensive to use an ordered map)

Reply:
Can do this in a later CL

Reply:
Done in this CL.